emmarin/cl/cl/src/crust/tx.rs

```diff
@@ -0,0 +1,334 @@
+use rand_core::{CryptoRngCore, RngCore};
+use serde::{Deserialize, Serialize};
+use std::collections::BTreeMap;
+
+use crate::{
+    crust::{
+        balance::{Balance, BalanceWitness},
+        iow::{BurnWitness, InputWitness, MintWitness, OutputWitness},
+        NoteCommitment, Nullifier,
+    },
+    ds::mmr::{MMRProof, MMR},
+    mantle::ZoneId,
+    merkle, Digest, Hash,
+};
+
+pub const MAX_INPUTS: usize = 8;
+pub const MAX_OUTPUTS: usize = 8;
+
+/// An identifier of a transaction
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Serialize, Deserialize)]
+pub struct TxRoot(pub [u8; 32]);
+
+/// An identifier of a bundle
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Serialize, Deserialize)]
+pub struct BundleRoot(pub [u8; 32]);
+
+impl From<[u8; 32]> for TxRoot {
+    fn from(bytes: [u8; 32]) -> Self {
+        Self(bytes)
+    }
+}
+
+impl From<[u8; 32]> for BundleRoot {
+    fn from(bytes: [u8; 32]) -> Self {
+        Self(bytes)
+    }
+}
+
+impl TxRoot {
+    pub fn random(mut rng: impl RngCore) -> Self {
+        let mut sk = [0u8; 32];
+        rng.fill_bytes(&mut sk);
+        Self(sk)
+    }
+
+    pub fn hex(&self) -> String {
+        hex::encode(self.0)
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct Tx {
+    pub root: TxRoot,
+    pub balance: Balance,
+    pub updates: Vec<LedgerUpdate>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct TxWitness {
+    pub inputs: Vec<InputWitness>,
+    pub outputs: Vec<(OutputWitness, Vec<u8>)>,
+    pub data: Vec<u8>,
+    pub balance_blinding: [u8; 16],
+    pub mint_burn_blinding: [u8; 16],
+    pub mints: Vec<MintWitness>,
+    pub burns: Vec<BurnWitness>,
+    pub frontier_paths: Vec<(MMR, MMRProof)>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct LedgerUpdate {
+    pub zone_id: ZoneId,
+    pub frontier_nodes: Vec<MMR>,
+    pub inputs: Vec<Nullifier>,
+    pub outputs: Vec<NoteCommitment>,
+}
+
+pub struct LedgerUpdateWitness {
+    pub zone_id: ZoneId,
+    pub frontier_nodes: Vec<MMR>,
+    pub inputs: Vec<Nullifier>,
+    pub outputs: Vec<(NoteCommitment, Vec<u8>)>,
+}
+
+impl LedgerUpdateWitness {
+    pub fn commit(self) -> (LedgerUpdate, [u8; 32]) {
+        let input_root = merkle::root(&merkle::padded_leaves(
+            &self.inputs.iter().map(|nf| nf.0).collect::<Vec<_>>(),
+        ));
```

Comment:
I'm wondering if there is any efficiency gains from the sparse tree optimizations we have in frontier root:

```rust
let input_root = frontier_root(MMR::from_iter(self.inputs.iter().map(|nf| nf.0)))
```

Reply:
right, maybe we don't need the non-sparse one anymore (except maybe as a reference test)